modules/opentracing/src/main/scala/OTTracer.scala

```diff
@@ -0,0 +1,54 @@
+// Copyright (c) 2019 by Rob Norris
+// This software is licensed under the MIT License (MIT).
+// For more information see LICENSE or https://opensource.org/licenses/MIT
+
+package natchez
+package opentracing
+
+import cats.effect.{Resource, Sync}
+import cats.syntax.applicative._
+import cats.syntax.applicativeError._
+import io.{opentracing => ot}
+import io.opentracing.propagation.{Format, TextMapAdapter}
+
+import scala.jdk.CollectionConverters._
+
+object OTTracer {
+
+  private[opentracing] def finishSpan[F[_] : Sync](t: ot.Tracer): ot.Span => F[Unit] = s =>
+    Sync[F].delay {
+      t.scopeManager().activate(s)
+      s.finish()
+    }
+
+  def entryPoint[F[_] : Sync](acquireTracer: F[ot.Tracer]): Resource[F, EntryPoint[F]] =
+    Resource.fromAutoCloseable(acquireTracer)
+      .map(t => new EntryPoint[F] {
+        override def root(name: String): Resource[F, Span[F]] = {
+          val acquireSpan: F[ot.Span] = Sync[F].delay(t.buildSpan(name).start())
+          Resource.make(acquireSpan)(finishSpan(t)).map(OTSpan(t, _))
+        }
+
+        override def continue(name: String, kernel: Kernel): Resource[F, Span[F]] = {
+          val acquireSpan: F[ot.Span] = Sync[F].delay {
+            val p = t.extract(
+              Format.Builtin.HTTP_HEADERS,
+              new TextMapAdapter(kernel.toHeaders.asJava)
+            )
+            t.buildSpan(name).asChildOf(p).start()
+          }
+
+          Resource.make(acquireSpan)(finishSpan(t)).map(OTSpan(t, _))
+        }
+
+        override def continueOrElseRoot(name: String, kernel: Kernel): Resource[F, Span[F]] =
+          continue(name, kernel) flatMap {
+            case null => root(name) // hurr, means headers are incomplete or invalid
+            case a    => a.pure[Resource[F, *]]
+          } recoverWith {
+            case _: Exception => root(name)
```

Comment:
This was a more specific exception in the Jaeger implementation. I'm not sure if widening it is a problem—if it is, I can let implementations pass in an exception handler (or something like that).